File: leetcode/binary_tree/is_balanced.rb
def max_depth(tree, cache)
  return 0 if tree.nil?
  return cache[tree] if cache[tree]

  left, right = tree.left, tree.right
  left_result = cache[left] || max_depth(left) + 1
  right_result = cache[right] || max_depth(right) + 1

  cache[tree] = [left_result, right_result].max
end

def is_balanced_helper(root, cache={})
  return true if root.nil?

  max_left = max_depth(root.left, cache)
  max_right = max_depth(root.right, cache)

  min, max = [max_left, max_right].sort

  if max - min <= 1
    return is_balanced(root.left, cache) && is_balanced(root.right, cache)
  else
    return false
  end
end

def is_balanced(root)
  is_balanced_helper(root, {})
end

File: leetcode/binary_tree/isBalanced.js
var maxDepth = function(root) {
  if (root === null) return 0;

  const left = root.left;
  const right = root.right;
  const left_result = maxDepth(left) + 1
  const right_result = maxDepth(right) + 1

  return Math.max(left_result, right_result);
};

var isBalanced = function(root) {
  if (root === null) return true;

  const left = root.left;
  const right = root.right;

  const left_d = maxDepth(left);
  const right_d = maxDepth(right);

  if (Math.abs(right_d - left_d) <= 1) return isBalanced(left) && isBalanced(right);

  return false;
};
